fish-rust/src/wutil/wcstod.rs

```diff
@@ -106,6 +111,94 @@
     }
 }
 
+/// Like [`wcstod()`], but allows underscore separators. Leading, trailing, and multiple underscores
+/// are allowed, as are underscores next to decimal (`.`), exponent (`E`/`e`/`P`/`p`), and
+/// hexadecimal (`X`/`x`) delimiters. This consumes trailing underscores -- `consumed` will include
+/// the last underscore which is legal to include in a parse (according to the above rules).
+/// Free-floating leading underscores (`"_ 3"`) are not allowed and will result in a no-parse.
+/// Underscores are not allowed before or inside of `"infinity"` or `"nan"` input. Trailing
+/// underscores after `"infinity"` or `"nan"` are not consumed.
+pub fn wcstod_underscores<Chars>(
+    s: Chars,
+    decimal_sep: char,
+    consumed: &mut usize,
+) -> Result<f64, Error>
+where
+    Chars: IntoCharIter,
+{
+    let mut chars = s.chars().peekable();
+
+    let mut leading_whitespace = 0;
+    // Skip leading whitespace.
+    while let Some(c) = chars.peek() {
+        if c.is_ascii_whitespace() {
+            leading_whitespace += 1;
+            chars.next();
+        } else {
+            break;
+        }
+    }
+
+    let is_sign = |c: char| "+-".contains(c);
+    let is_inf_or_nan_char = |c: char| "iInN".contains(c);
+
+    // We don't do any underscore-stripping for infinity/NaN.
+    let mut is_inf_nan = false;
+    if let Some(&c1) = chars.peek() {
+        if is_inf_or_nan_char(c1) {
+            is_inf_nan = true;
+        } else if is_sign(c1) {
+            // FIXME make this more efficient
+            let mut copy = chars.clone();
```

Comment:
sounds like it was maybe a bit premature to take an iterator instead of a char slice like in C++.
No big deal either way, this is fine.